### DISCUSSION TITLE
监视视频cc字幕的镜像元素的频繁变化并进行翻译

### DISCUSSION THREAD

**Discussion by Oshibuki** [immersive-translate/immersive-translate#419] [Ideas]
### 新功能 | New Feature

1.监视特定元素的频繁变化
2.在每次元素的文本变化时进行翻译

### 为什么您想要添加这些功能 | Why you want to add these features

1.某些视频网站只提供了某种语言的cc字幕，这些字幕无法被翻译
2.这种页面的结构往往是标准的`<video><source /><track /></video>`
3.通过以下代码可以将当前播放的字幕提取到一个新元素来展示：
```
video.addEventListener('timeupdate', e => {
      const track = [...video.textTracks]
        .find(track => track.mode === "showing");
      const texts = [...track.activeCues].map(cue => cue.text);
      let newVal = texts.join('\n');
      let currentVal = document.getElementById('currentSub').innerHTML;
      if (newVal !== currentVal) {
        setTimeout(()=>{document.getElementById('currentSub').innerHTML = newVal;},500)
      }
    });
```
4.目前只能翻译一次新元素的文本，对于后续变化无法追踪

### 提议 | Proposal

选择特定元素添加监听器侦测变化，变化发生时触发翻译

**Comment by theowenyoung** [comment]
这个应该不是所有视频网站都一样吧？

可以给个网址吗？

**Reply by Oshibuki** [reply]
主要是为了看类似coursera之类站点的技术视频，这些网站往往只有英文的cc字幕，但是结构应该是很合规的
https://learn.deeplearning.ai/chatgpt-prompt-eng/lesson/1/introduction
currentSub是我自己添加的div元素

**Reply by Oshibuki** [reply]
> 这个应该不是所有视频网站都一样吧？
> 
> 可以给个网址吗？

我写了个简单的demo脚本，https://greasyfork.org/zh-CN/scripts/465174-cc-caption-automatic-translator  希望能起到参考作用
